### PR TITLE
Run2-hgx192 Complete first level of debugging geometry

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Complete first level of debugging V10 geometry of HGCal. There was a bug in finding the layer # from position (affecting only the last layer of the group #28 for EE and #24 or 22 for HE for V9/V10 version). This will surely affect getClosestCell(GlobalPoint) method and applicable to V10 as well as V9 versions.

#### PR validation:

Tested with the two workflows 24034.0 (D28 geometry) and 29034.0 (D41 geometry)

#### if this PR is a backport please specify the original PR:

The changes are bug removals and hence applicable to pre3 and pre4 of 10_6_0

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc [hunk 1]
```diff
@@ -387,14 +387,14 @@ int HGCalDDDConstants::getLayer(double z, bool reco) const {
...
-                 ++k;
-                 return zz < 0.5 * (zLayerHex[k - 1] + zLayerHex[k]);
-               });
-  int lay = k;
+  auto itr = std::find_if(zLayerHex.begin() + 1, zLayerHex.end(),
+			  [&k, &zz, &zLayerHex](double zLayer) {
+			    ++k;
+			    return zz < 0.5 * (zLayerHex[k - 1] + zLayerHex[k]);
+			  });
+  int lay =(itr == zLayerHex.end()) ? (int)(zLayerHex.size()) :  k;
```
Comment (cvuosalo) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]:
```suggestion
  int lay =(itr == zLayerHex.end()) ? static_cast<int>(zLayerHex.size()) :  k;
```

Comment (cvuosalo) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: C++ style casts are preferred by the coding [rules](https://github.com/cms-sw/cms-sw.github.io/blob/3e89200bde531a9aa1d01d4c92c1d65a5f6c87fd/cms_coding_rules.md#4----technical-coding-rules-1).

OK to make this change in a later PR.

**Comment by cvuosalo** [comment]
+1

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by cvuosalo** [comment]
+1

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc (hunk 1 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: Done

**Comment by kpedro88** [comment]
@bsunanda please clarify exactly which bugs were fixed. I see comparison differences not just in D41, but also in D35, which was used for the MTD TDR and in principle should not be affected by changes to the HGCal v10 geometry.

**Comment by bsunanda** [comment]
@kpedro88 Yes there was a bug in getCLosestCell which will affect V9 as well as V10 geometry. This is expected.

**Comment by bsunanda** [comment]
The bug was affecting the last layer of a given setup (layer 28 for EE and layer# 24 or 22 for HE depending on the HGCal version #). I prefer to have bug corrected code in the repository. MTD production was made with earlier CMSSW version and they can continue with that version for MTD analysis.

**Comment by kpedro88** [comment]
@bsunanda thanks, this is acceptable as long as there is a legitimate, understood reason for the change. Can you update the PR description with these details?

**Comment by civanch** [comment]
+1

**Comment by kpedro88** [comment]
+upgrade

**Comment by kpedro88** [comment]
+1

**Pull request merged**